cmd/reaper/resource_types.go

```diff
@@ -0,0 +1,108 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/boskos/common"
+)
+
+type ResourceTypes interface {
+	Types() []string
+}
+
+type resourceTypesStatic struct {
+	providedTypes []string
+}
+
+func (rts *resourceTypesStatic) Types() []string { return rts.providedTypes }
+
+type resourceTypesWatcher struct {
+	resourceTypesStatic
+	config      string
+	mergedTypes []string
+	rw          sync.RWMutex
+}
+
+func NewResourceTypes(providedTypes []string, config string) (ResourceTypes, error) {
+	if config == "" {
+		return &resourceTypesStatic{providedTypes: providedTypes}, nil
+	}
+
+	rtw := resourceTypesWatcher{
+		resourceTypesStatic: resourceTypesStatic{
```

Comment:
why not hold on to `sets.NewString(providedTypes)` instead of these layers of indirection?

Reply:
I'm using a `sets.String` as a temporary, transient area of storage because it's a convenient mean to merge the `types` from both the config file and the command line.
I won't store the set as a member of `ResourceTypes` struct cause it would be difficult to recompute the `types` when the config file changes.
Let's picture the following use case:
- `providedTypes` is `{"a", "b"}`
- `types` from config are `{"b", "d"}`
- the resulting `types` would be `{"a", "b", "d"}`
- now the config changes into `{"b", "e"}`
- at this stage would be not so obvious to get the correct result `{"a", "b", "e"}` starting only from the last known set `{"a", "b", "d"}`. It would be much easier just to `sets.NewString(providedTypes...).Insert(newConfigTypes...).List()`

I'm obviously open to changes if you see something that I don't.

Reply:
Store the set of `providedTypes`, not the merged set.

Reply:
Is this (or the like) what you mean?
```golang
func (rtw *resourceTypesWatcher) Types() []string {
	types, err := rtw.mergeTypesFromConfig()
	if err != nil {
		logrus.Warnf("merge types: %s", err.Error())
		return rtw.providedTypes
	}
	return types
}
```
That is: reading from the file and therefore merging with the provided ones without bothering at all with the "watcher stuff"?

Reply:
No - I am saying just that the whole `resourceTypesStatic` type is superfluous - you could just as well write

```go
type resourceTypesWatcher struct {
	providedTypes sets.String
	config      string
	mergedTypes []string
	rw          sync.RWMutex
}
```

Reply:
Removed, it looks like yours.